README.md
# clj-openweather-tray

System tray app for Windows OS written in Clojure that shows outside temperature in given city 
using data from openweathermap.org. 
Color of text is changing based on the temperature.
Supports both Celsius and Fahrenheit scale.

Tooltip contains additional weather info (wind speed, direction, sunrise, sunset, etc.)

Screenshot:
<img src="images/screenshot.png">

## Motivation

I created this application for two reasons:
1. To learn Clojure Java interop.
2. I'm using similar application but it displays cloud/sun/whatever icons instead of outside temperature:
https://github.com/FelixdelasPozas/TrayWeather. I recommend it if you need more functionality.

## Build

    lein uberjar

## Installation

1. Copy generated standalone jar file somewhere along with the config file and bat file, e.g:
    
        mkdir somedir
        cp .\target\uberjar\clj-openweather-tray-0.0.4-standalone.jar somedir\
        cp .\clj-openweather-tray-conf.edn somedir\
        cp .\clj-openweather-tray.bat somedir\
    
2. Open config file and add city-id and api-key.

* You can get list of city ids here: http://bulk.openweathermap.org/sample/.
* Follow these instructions to get your api key: https://openweathermap.org/api.

## Usage

Run bat file:

    .\clj-openweather-tray.bat

## TODO
Create binary using GraalVM's native-image. Swing support is still missing: https://github.com/oracle/graal/issues/1327.

## Findings
16 pixels, which is the size of tray icon on Windows OS, is not much if you want to fit 3 or 4 chars inside and still
be able to read it :)

## License

Copyright © 2020 Miroslav Havrlent

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [0.0.4] - 2020-03-28
### Fixed
- Fix NPE in deg_to_cardinal function caused by missing wind degrees property in json.

### Added
- native-image config for lein native-image plugin.

## [0.0.3] - 2020-03-27
### Added
- Tooltip with additional weather info.

### Changed
- Updated README.md and CHANGELOG.md with new tooltip info.
- Added screenshot of tray.

## [0.0.2] - 2020-03-25
### Added
- This changelog.

### Changed
- Change config file format from json to edn to be able to define colors as numbers.
- Update README.md.

### Fixed
- Fix bug in find-closest function to return closest number event for key k outside of sorted-map sm
- Fix icon-size.

test/clj_openweather_tray/core_test.clj
(ns clj-openweather-tray.core-test
  (:import [java.awt SystemTray Color])
  (:require [clojure.test :refer :all]
            [clj-openweather-tray.core :refer :all]))

(deftest create-tray-icon-test
  (testing "tray icon"
    (let [tray-icon (create-tray-icon "5" Color/WHITE "Arial" 0 11 16 "tooltip text")
          img (.getImage tray-icon)
          img-height (.getHeight img)
          img-width (.getWidth img)
          popup (.getPopupMenu tray-icon)
          item (.getItem popup 0)
          item-label (.getLabel item)]
      (testing "image height"
        (is (= 16 img-height)))
      (testing "image width"
        (is (= 16 img-width)))
      (testing "exit item label"
        (is (= "Exit" item-label))))))

(deftest add-icon-to-tray-test
  (testing "add icon to tray"
    (add-icon-to-tray (create-tray-icon "5" Color/WHITE "Arial" 0 11 16 "tooltip text"))
    (let [tray (SystemTray/getSystemTray)
          tray-icons (.getTrayIcons tray)]
      (testing "number of icons in tray should be one"
        (is (= 1 (count tray-icons)))))))

(deftest remove-all-icons-from-tray-test
  (testing "remove all icons from tray"
    (add-icon-to-tray (create-tray-icon "5" Color/WHITE "Arial" 0 11 16 "tooltip text")))
  (remove-all-icons-from-tray)
  (let [tray (SystemTray/getSystemTray)
        tray-icons (.getTrayIcons tray)]
    (testing "number of icons in tray should be zero"
      (is (= 0 (count tray-icons))))))

(deftest update-tray-icon-test
  (testing "update tray icon"
    (update-tray-icon "5" Color/WHITE "Arial" 0 11 16 "tooltip text")
    (let [tray (SystemTray/getSystemTray)
          tray-icons (.getTrayIcons tray)]
      (testing "number of icons in tray should be one"
        (is (= 1 (count tray-icons)))))))

(deftest get-weather-data-test
  (testing "get temperature"
    (testing "using sample weather API"
      (is (= 300.15 (-> (get-weather-data
                          "https://samples.openweathermap.org/data/2.5/weather?id=%s&appid=%s"
                          "2172797"
                          "b6907d289e10d714a6e88b30761fae22")
                        (:main)
                        (:temp)))))))

(deftest kelvin-to-celsius-test
  (testing "kelvin to celsius conversion"
    (testing "1 kelvin to celsius"
      (is (= -272.15 (kelvin-to-celsius 1))))
    (testing "273.15 kelvin to celsius"
      (is (= 0.0 (kelvin-to-celsius 273.15))))))

(deftest kelvin-to-fahrenheit-test
  (testing "kelvin to fahrenheit conversion"
    (testing "1 kelvin to fahrenheit"
      (is (= -457.8714388489209 (kelvin-to-fahrenheit 1))))
    (testing "273.15 kelvin to fahrenheit"
      (is (= 31.606978417266077 (kelvin-to-fahrenheit 273.15))))))

(deftest convert-temperature-test
  (testing "testing temperature conversion using"
    (testing "celsius scale"
      (is (= -272.15 (convert-temperature 1 "C"))))
    (testing "fahrenheit scale"
      (is (= -457.8714388489209 (convert-temperature 1 "F"))))))

(deftest abs-test
  (testing "absolute value of 0 should be 0"
    (is (= 0 (abs 0))))
  (testing "absolute value of 10 should be 10"
    (is (= 10 (abs 10))))
  (testing "absolute value of -10 should be 10"
    (is (= 10 (abs -10)))))

(deftest find-closest-test
  (testing "find closest key in sorted-map"
    (let [sm (sorted-map 9999 "c" 1 "a" 500 "b")]
      (do
        (is (= 500 (find-closest sm 1000)))
        (is (= 1 (find-closest sm 1)))
        (is (= 1 (find-closest sm -0)))
        (is (= 1 (find-closest sm 0)))
        (is (= 1 (find-closest sm 250)))
        (is (= 9999 (find-closest sm 10000)))
        (is (= 1 (find-closest sm -1000)))))))

(deftest get-color-test
  (testing "get white color for 273 K"
    (let [color (get-color 273 {273 "#ffffff"})]
      (do
        (is (= 255 (.getRed color)))
        (is (= 255 (.getGreen color)))
        (is (= 255 (.getBlue color)))))))

(deftest deg-to-cardinal-test
  (testing "convert degrees to cardinal point"
    (do
      (is (= "N" (deg-to-cardinal 0)))
      (is (= "NNE" (deg-to-cardinal 25)))
      (is (= "NE" (deg-to-cardinal 50)))
      (is (= "ENE" (deg-to-cardinal 75)))
      (is (= "E" (deg-to-cardinal 100)))
      (is (= "ESE" (deg-to-cardinal 110)))
      (is (= "SE" (deg-to-cardinal 125)))
      (is (= "SSE" (deg-to-cardinal 150)))
      (is (= "S" (deg-to-cardinal 175)))
      (is (= "SSW" (deg-to-cardinal 200)))
      (is (= "SW" (deg-to-cardinal 225)))
      (is (= "WSW" (deg-to-cardinal 250)))
      (is (= "W" (deg-to-cardinal 275)))
      (is (= "WNW" (deg-to-cardinal 300)))
      (is (= "NW" (deg-to-cardinal 325)))
      (is (= "NNW" (deg-to-cardinal 335)))
      (is (= "N" (deg-to-cardinal 350)))
      (is (= "N" (deg-to-cardinal 360)))
      (is (= "N" (deg-to-cardinal 365)))
      (is (= "NNE" (deg-to-cardinal 385))))))

(deftest clouds-percentage-to-text-test
  (testing "5% should be clear sky"
    (is (= "Clear sky" (clouds-percentage-to-text 5))))
  (testing "86% should be overcast clouds"
    (is (= "Overcast clouds" (clouds-percentage-to-text 86)))))

(deftest extract-hour-and-minute-from-epoch-test
  (testing "1585283921 should return 4:38 UTC"
    (is (= "4:38 UTC" (extract-hour-and-minute-from-epoch 1585283921)))))

project.clj
(defproject mhavrlent/clj-openweather-tray "0.0.4"
  :description "System tray app for Windows OS that shows outside temperature in given city using data from openweathermap.org."
  :url "https://github.com/mhavrlent/clj-openweather-tray"
  :license {:name "EPL-2.0 OR GPL-2.0-or-later WITH Classpath-exception-2.0"
            :url  "https://www.eclipse.org/legal/epl-2.0/"}
  :dependencies [[org.clojure/clojure "1.10.1"]
                 [clj-http "3.10.0"]
                 [org.clojure/data.json "1.0.0"]]
  :main ^:skip-aot clj-openweather-tray.core
  :target-path "target/%s"
  :profiles {:uberjar {:aot :all}}
  :native-image {:opts ["--verbose"
                        "--no-fallback"
                        "--no-server"
                        "--report-unsupported-elements-at-runtime"
                        "-H:+ReportExceptionStackTraces"
                        "-J-Dclojure.spec.skip-macros=true"
                        "-J-Dclojure.compiler.direct-linking=true"
                        "-H:ReflectionConfigurationFiles=reflection.json"
                        "--initialize-at-build-time"
                        "-J-Xmx3g"
                        "-H:Log=registerResource:"]})


src/clj_openweather_tray/core.clj
(ns clj-openweather-tray.core
  (:gen-class)
  (:import [java.awt SystemTray TrayIcon PopupMenu MenuItem Font Color]
           [java.awt.event ActionListener]
           (javax.swing UIManager)
           (java.awt.image BufferedImage)
           (java.time Instant ZonedDateTime ZoneOffset))
  (:require [clojure.string :as str]
            [clj-http.client :as client]
            [clojure.data.json :as json]
            [clojure.edn :as edn]))

(defn create-tray-icon
  "Creates awt TrayIcon with text instead of icon and Exit MenuItem to close the application."
  [text color font font-style font-size icon-size tooltip-text]
  (let [popup (PopupMenu.)
        font (Font. font font-style font-size)
        img (BufferedImage. icon-size icon-size BufferedImage/TYPE_INT_ARGB)
        g2d (.createGraphics img)
        _ (.setFont g2d font)
        _ (.setColor g2d color)
        x (if (< 2 (count text)) -1 1)
        _ (.drawString g2d text x (+ 1 font-size))
        _ (.dispose g2d)
        tray-icon (TrayIcon. img)
        tray (SystemTray/getSystemTray)
        exit-item (MenuItem. "Exit")
        _ (.addActionListener exit-item (reify ActionListener
                                          (actionPerformed [_ _]
                                            (do (.remove tray tray-icon)
                                                (System/exit 0)))))
        _ (.add popup exit-item)
        _ (.setPopupMenu tray-icon popup)
        _ (.setToolTip tray-icon tooltip-text)]
    tray-icon))

(defn remove-all-icons-from-tray
  "Removes all icons from the tray."
  []
  (let [tray (SystemTray/getSystemTray)
        tray-icons (.getTrayIcons tray)]
    (doseq [ti tray-icons]
      (.remove tray ti))))

(defn add-icon-to-tray
  "Adds icon to the tray."
  [icon]
  (let [tray (SystemTray/getSystemTray)]
    (.add tray icon)))

(defn update-tray-icon
  "Updates tray icon."
  [text color font font-style font-size icon-size tooltip-text]
  (remove-all-icons-from-tray)
  (add-icon-to-tray (create-tray-icon text color font font-style font-size icon-size tooltip-text)))

(defn get-weather-data
  "Gets current weather data for given city id using api key from openweathermap.org.
  List of city IDs can be downloaded here http://bulk.openweathermap.org/sample/."
  [url city-id api-key]
  (let [url (format url city-id api-key)]
    (-> (client/get url)
        (:body)
        (json/read-str :key-fn keyword))))

(defn kelvin-to-celsius
  "Converts Kelvin to Celsius"
  [kelvin]
  (- kelvin 273.15))

(defn kelvin-to-fahrenheit
  "Converts Kelvin to Fahrenheit"
  [kelvin]
  (- (/ kelvin 0.556) 459.67))

(defn convert-temperature
  "Converts given temperature in Kelvin to defined scale, which can be C for Celsius or F for Fahrenheit."
  [temp scale]
  (cond
    (= (str/upper-case scale) "C") (kelvin-to-celsius temp)
    (= (str/upper-case scale) "F") (kelvin-to-fahrenheit temp)))

(defn abs
  "Returns absolute value of x."
  [x]
  (if (neg? x) (- x) x))

(defn find-closest
  "Finds closest key in a sorted map sm using provided key k."
  [sm k]
  (let [a (subseq sm <= k)
        ak (if (empty? a) nil (key (last a)))
        b (subseq sm >= k)
        bk (if (empty? b) nil (key (first b)))]
    (if (= ak k)
      k
      (if (= bk k)
        k
        (if (and (some? ak) (some? bk))
          (if (< (abs (- k bk)) (abs (- k ak)))
            bk
            ak)
          (if (some? ak)
            ak
            (if (some? bk)
              bk)))))))

(defn get-color
  "Returns color based on temperature. The higher the temperature, the warmer the color."
  [temp colors]
  (let [sm (into (sorted-map) colors)
        closest-color-key (find-closest sm (Math/round (.doubleValue temp)))
        closest-color-value (get sm closest-color-key)]
    (Color/decode closest-color-value)))

(defn deg-to-cardinal
  "Converts degrees to cardinal point."
  [deg]
  (if (some? deg)
    (let [points ["N", "NNE", "NE", "ENE", "E", "ESE", "SE", "SSE", "S", "SSW", "SW", "WSW", "W", "WNW", "NW", "NNW"]
          points-count (count points)
          index (Math/round (.doubleValue (/ deg (/ 360 points-count))))]
      (get points (mod index points-count)))
    "Data missing"))

(defn clouds-percentage-to-text
  "Converts percentage of clouds to text."
  [perc]
  (cond
    (and (>= perc 0) (<= perc 10)) "Clear sky"
    (and (>= perc 11) (<= perc 24)) "Few clouds"
    (and (>= perc 25) (<= perc 50)) "Scattered clouds"
    (and (>= perc 51) (<= perc 84)) "Broken clouds"
    (and (>= perc 85) (<= perc 100)) "Overcast clouds"))

(defn extract-hour-and-minute-from-epoch
  "Converts epoch time to human readable string."
  [epoch]
  (let [instant (Instant/ofEpochSecond epoch)
        zoned-date-time (ZonedDateTime/ofInstant instant (ZoneOffset/UTC))
        hour (.getHour zoned-date-time)
        minute (.getMinute zoned-date-time)]
    (format "%s:%s UTC" hour minute)))

(defn create-tooltip-text
  "Creates text for the tray tooltip that contains additional weather information."
  [wind-speed wind-deg humidity clouds sunrise sunset]
  (format "Wind speed: %s m/s\nWind direction: %s (%s)\nHumidity: %s %%\nSky: %s\nSunrise: %s\nSunset: %s"
          wind-speed
          wind-deg
          (deg-to-cardinal wind-deg)
          humidity
          (clouds-percentage-to-text clouds)
          (extract-hour-and-minute-from-epoch sunrise)
          (extract-hour-and-minute-from-epoch sunset)))

(defn -main
  "Main method. Updates the tray with temperature from openweathermap.org and then sleeps for x seconds
  (see sleep-interval in config file)."
  [& args]
  (when-not (. SystemTray isSupported)
    (throw (Exception. "SystemTray not supported on this platform.")))

  (while true
    (let [config (edn/read-string (slurp "clj-openweather-tray-conf.edn"))
          weather-data (get-weather-data (:url config)
                                         (:city-id config)
                                         (:api-key config))
          temp (-> weather-data (:main) (:temp))
          converted-temp (convert-temperature temp (:scale config))
          color (get-color temp (:colors config))
          wind-speed (-> weather-data (:wind) (:speed))
          wind-deg (-> weather-data (:wind) (:deg))
          humidity (-> weather-data (:main) (:humidity))
          clouds (-> weather-data (:clouds) (:all))
          sunrise (-> weather-data (:sys) (:sunrise))
          sunset (-> weather-data (:sys) (:sunset))]
      (do
        (UIManager/setLookAndFeel (:look-and-feel config))
        (update-tray-icon (str (Math/round (.doubleValue converted-temp)))
                          color
                          (:font config)
                          (:font-style config)
                          (:font-size config)
                          (:icon-size config)
                          (create-tooltip-text wind-speed wind-deg humidity clouds sunrise sunset))
        (Thread/sleep (:sleep-interval config))))))